HeterogeneousCore/SonicTriton/src/TritonClient.cc

```diff
@@ -201,76 +220,119 @@ void TritonClient::evaluate() {
-
-              if (verbose()) {
-                const auto& stats = summarizeServerStats(start_status, end_status);
-                reportServerSideStats(stats);
-              }
-
-              //check result
-              status = getResults(results_ptr);
-
-              //finish
-              finish(status);
-            },
-            options_,
-            inputsTriton_,
-            outputsTriton_),
-        "evaluate(): unable to launch async run");
-
-    //if AsyncRun failed, finish() wasn't called
-    if (!status)
-      finish(false);
+    try {
+      triton_utils::throwIfError(
+          client_->AsyncInfer(
+              [t1, start_status, this](nic::InferResult* results) {
+                //get results
+                std::shared_ptr<nic::InferResult> results_ptr(results);
+                try {
+                  triton_utils::throwIfError(results_ptr->RequestStatus(), "evaluate(): unable to get result");
+                } catch (...) {
+                  if (!checkEptr())
+                    return;
+                }
+                auto t2 = std::chrono::high_resolution_clock::now();
+
+                if (!debugName_.empty())
+                  edm::LogInfo(fullDebugName_)
+                      << "Remote time: " << std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
+
+                if (verbose()) {
+                  inference::ModelStatistics end_status;
+                  try {
+                    end_status = getServerSideStatus();
+                  } catch (...) {
+                    if (!checkEptr())
+                      return;
+                  }
+
+                  const auto& stats = summarizeServerStats(start_status, end_status);
+                  reportServerSideStats(stats);
+                }
+
+                //check result
+                try {
+                  getResults(results_ptr);
+                } catch (...) {
+                  if (!checkEptr())
+                    return;
+                }
+
+                //finish
+                finish(true);
+              },
+              options_,
+              inputsTriton_,
+              outputsTriton_),
+          "evaluate(): unable to launch async run");
+    } catch (...) {
+      //if AsyncRun failed, finish() wasn't called
+      if (!checkEptr())
+        return;
+    }
   } else {
     //blocking call
     auto t1 = std::chrono::high_resolution_clock::now();
     nic::InferResult* results;
-    bool status = triton_utils::warnIfError(client_->Infer(&results, options_, inputsTriton_, outputsTriton_),
-                                            "evaluate(): unable to run and/or get result");
-    if (!status) {
-      finish(false);
-      return;
+    try {
+      triton_utils::throwIfError(client_->Infer(&results, options_, inputsTriton_, outputsTriton_),
+                                 "evaluate(): unable to run and/or get result");
+    } catch (...) {

```

Comment:
In all `catch(...)` please annotate the `try` with `CMS_SA_ALLOW` and add a comment along
```suggestion
    // Caught exception is propagated via WaitingTaskWithArenaHolder
    CMS_SA_ALLOW try {
      triton_utils::throwIfError(client_->Infer(&results, options_, inputsTriton_, outputsTriton_),
                                 "evaluate(): unable to run and/or get result");
    } catch (...) {

```
to silence static analyzer warnings.

An alternative, when you know the exact exception (base class) type that will be thrown, would be to catch that type explicitly.

Reply:
I only use `cms::Exception` in SONIC packages; however, `cudaCheck()` currently throws `std::runtime_error`. Is it permissible to change `cudaCheck()` to throw a `cms::Exception`? If not, I should probably use `CMS_SA_ALLOW`.

Reply:
No, I'd rather not make `cudaCheck()` CMS specific.

Reply:
They both derive from `std::exception` so you could catch that.

Reply:
I realized that in 7b9a0ca53d97c781459e48bfe94eae3125acbb23, while eagerly pursuing simplifications, the ability to retry after some possibly-transient exceptions from the server was lost. I've restored that in the latest commits by adding a `TritonException` class that denotes those exceptions and can be handled differently. (I tried two approaches to reduce duplication: preprocessor macro and function templates/lambdas. The former has the nice property that it can also handle the `return` statement, but I don't think the `CMS_SA_ALLOW` will be recognized in that form, so I ultimately chose the latter.)